src/node/historical_queries.h

```diff
@@ -0,0 +1,341 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the Apache 2.0 License.
+#pragma once
+
+#include "consensus/ledger_enclave_types.h"
+#include "kv/store.h"
+#include "node/historical_queries_interface.h"
+#include "node/history.h"
+#include "node/rpc/node_interface.h"
+
+#include <deque>
+#include <map>
+#include <memory>
+
+namespace ccf::historical
+{
+  class StateCache : public AbstractStateCache
+  {
+  protected:
+    kv::Store& source_store;
+    ringbuffer::WriterPtr to_host;
+
+    enum class RequestStage
+    {
+      Fetching,
+      Untrusted,
+      Trusted,
+    };
+
+    using LedgerEntry = std::vector<uint8_t>;
+
+    struct Request
+    {
+      RequestStage current_stage = RequestStage::Fetching;
+      crypto::Sha256Hash entry_hash = {};
+      StorePtr store = nullptr;
+    };
+
+    // TODO: Make this an LRU, evict old entries
+    static constexpr size_t MAX_ACTIVE_REQUESTS = 10;
+    std::map<consensus::Index, Request> requests;
+
+    // To trust an index, we currently need to fetch a sequence of entries
+    // around it - these aren't requests, so we don't store them, but we do
+    // need to distinguish things-we-asked-for from junk-from-the-host
+    std::set<consensus::Index> pending_fetches;
+
+    void request_entry_at(consensus::Index idx)
+    {
+      if (requests.size() < MAX_ACTIVE_REQUESTS)
+      {
+        const auto ib = requests.insert(std::make_pair(idx, Request{}));
+        if (ib.second)
+        {
+          fetch_entry_at(idx);
+        }
+      }
+
+      // Too many outstanding fetches, this one is silently ignored
+      // TODO
+    }
+
+    void fetch_entry_at(consensus::Index idx)
+    {
+      const auto it =
+        std::find(pending_fetches.begin(), pending_fetches.end(), idx);
+      if (it != pending_fetches.end())
+      {
+        // Already fetching this index
+        return;
+      }
+
+      RINGBUFFER_WRITE_MESSAGE(
+        consensus::ledger_get,
+        to_host,
+        idx,
+        consensus::LedgerRequestPurpose::HistoricalQuery);
+      pending_fetches.insert(idx);
+    }
+
+    std::optional<ccf::Signature> get_signature(const StorePtr& sig_store)
+    {
+      kv::Tx tx;
+      auto sig_table = sig_store->get<ccf::Signatures>(ccf::Tables::SIGNATURES);
+      if (sig_table == nullptr)
+      {
+        throw std::logic_error(
```

Comment:
ccf_logic_error will also capture the callstack

Reply:
Thanks. We don't need the callstack here - there's a single caller, and we're going to catch and flatten the exception a short way up the stack. But I'll keep it in mind for future.